### PR TITLE
Improvements to type-related classes

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
This PR introduces additional improvements to working with type-related classes. In particular:
  * `KnownTypes` transformed into `enum`.
  * `Type`, `TypeSet` were made `@Immutable`

### CODE REVIEW

**Comment by alexander-yevsyukov** [comment]
@armiol, PTAL.

**Comment by alexander-yevsyukov** [comment]
@armiol, PTAL.

File: base/src/main/java/io/spine/type/KnownTypes.java [hunk 1]
```diff
@@ -45,15 +47,34 @@
...
  */
 @Internal
-public class KnownTypes {
+public enum KnownTypes {
 
-    private final TypeSet types;
+    INSTANCE;
+
+    @SuppressWarnings({
+            "TransientFieldNotInitialized", // We substitute the instance in readResolve()
```
Comment (armiol) [inline base/src/main/java/io/spine/type/KnownTypes.java]: As long as you have period at the end of the following sentence, let's have a period here as well.

File: testlib/src/test/java/io/spine/testing/VerifyShould.java [hunk 2]
```diff
@@ -1236,18 +1181,16 @@ public void run() {
...
+            throw runtimeException;
         };
 
         assertThrowsWithCause(runtimeException.getClass(), cause.getClass(),
                                      throwsRuntimeException);
     }
 
-    @SuppressWarnings("EqualsAndHashcode")
+    @SuppressWarnings({"EqualsAndHashcode", "EqualsHashCode"})
+        // Not having hashCode() is a part of this test. environment class.
```
Comment (armiol) [inline testlib/src/test/java/io/spine/testing/VerifyShould.java]: `test. environment class.` seem to have an extra period.

File: testlib/src/test/java/io/spine/testing/VerifyShould.java [hunk 3]
```diff
@@ -1236,18 +1181,16 @@ public void run() {
...
+        Runnable throwsRuntimeException = () -> {
+            throw runtimeException;
         };
 
         assertThrowsWithCause(runtimeException.getClass(), cause.getClass(),
                                      throwsRuntimeException);
     }
 
-    @SuppressWarnings("EqualsAndHashcode")
+    @SuppressWarnings({"EqualsAndHashcode", "EqualsHashCode"})
```
Comment (armiol) [inline testlib/src/test/java/io/spine/testing/VerifyShould.java]: I have looked at the issue, and it seems that there is no good way to eliminate this duplication. For some reason JetBrains and ErrorProne have different notion, and it seems non-customizable: in ErrorProne the warning name is [a part of their source code](https://github.com/google/error-prone/blob/master/core/src/main/java/com/google/errorprone/bugpatterns/EqualsHashCode.java#L60), and in IDEA I have found no option to override warning names as well.

So the only way is to have `hashCode()` :)

**Pull request merged**